cpp/src/arima/arima_helpers.cuh

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_runtime.h>
+
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/counting_iterator.h>
+
+#include "arima_common.h"
+
+namespace ML {
+
+/**
+ * Auxiliary function of reduced_polynomial. Computes a coefficient of an (S)AR
+ * or (S)MA polynomial based on the values of the corresponding parameters
+ *
+ * @tparam     isAr    Is this an AR (true) or MA (false) polynomial?
+ * @param[in]  param   Parameter array
+ * @param[in]  lags    Number of parameters
+ * @param[in]  idx     Which coefficient to compute
+ * @return             The value of the coefficient
+ */
+template <bool isAr>
+static inline __host__ __device__ double _param_to_poly(const double* param,
+                                                        int lags, int idx) {
+  if (idx > lags) {
+    return 0.0;
+  } else if (idx) {
+    return isAr ? -param[idx - 1] : param[idx - 1];
+  } else
+    return 1.0;
+}
+
+/**
+ * Helper function to compute the reduced AR or MA polynomial based on the
+ * AR and SAR or MA and SMA parameters
+ *
+ * @tparam     isAr    Is this an AR (true) or MA (false) polynomial?
+ * @param[in]  bid     Batch id
+ * @param[in]  param   Non-seasonal parameters
+ * @param[in]  lags    Number of non-seasonal parameters
+ * @param[in]  sparam  Seasonal parameters
+ * @param[in]  slags   Number of seasonal parameters
+ * @param[in]  s       Seasonal period
+ * @param[in]  idx     Which coefficient to compute
+ * @return             The value of the coefficient
+ */
+template <bool isAr>
+static inline __host__ __device__ double reduced_polynomial(
```

Comment:
Maybe because I initially didn't write the function as templated and `inline` can be ignored by the compiler when disabling the optimizations, so there is no guarantee that you won't have a double definition of the function. Templated functions don't have risk of double definition though so it's safe to remove this keyword.